tests/src/python/test_provider_hana.py

```diff
@@ -0,0 +1,361 @@
+                     '"id" INTEGER NOT NULL PRIMARY KEY,' \
+                     '"fld1" BOOLEAN)'
+        insert_sql = 'INSERT INTO "qgis_test"."boolean_type" ("id", "fld1") VALUES (?, ?)'
+        insert_args = [[1, 'TRUE'], [2, 'FALSE'], [3, None]]
+        self.prepareTestTable('boolean_type', create_sql, insert_sql, insert_args)
+
+        vl = self.createVectorLayer('table="qgis_test"."boolean_type" sql=', 'testbool')
+        self.assertTrue(vl.isValid())
+
+        fields = vl.dataProvider().fields()
+        self.assertEqual(fields.at(fields.indexFromName('fld1')).type(), QVariant.Bool)
+
+        values = {feat['id']: feat['fld1'] for feat in vl.getFeatures()}
+        expected = {1: True, 2: False, 3: NULL}
+        self.assertEqual(values, expected)
+
+    def testDateTimeTypes(self):
+        create_sql = 'CREATE TABLE "qgis_test"."date_time_type" ( ' \
+                     '"id" INTEGER NOT NULL PRIMARY KEY,' \
+                     '"date_field" DATE,' \
+                     '"time_field" TIME,' \
+                     '"datetime_field" TIMESTAMP)'
+        insert_sql = 'INSERT INTO "qgis_test"."date_time_type" ("id", "date_field", "time_field", "datetime_field") ' \
+                     'VALUES (?, ?, ?, ?)'
+        insert_args = [[1, '2004-03-04', '13:41:52', '2004-03-04 13:41:52']]
+        self.prepareTestTable('date_time_type', create_sql, insert_sql, insert_args)
+
+        vl = self.createVectorLayer('table="qgis_test"."date_time_type" sql=', 'testdatetimes')
+        self.assertTrue(vl.isValid())
+
+        fields = vl.dataProvider().fields()
+        self.assertEqual(fields.at(fields.indexFromName('date_field')).type(), QVariant.Date)
+        self.assertEqual(fields.at(fields.indexFromName('time_field')).type(), QVariant.Time)
+        self.assertEqual(fields.at(fields.indexFromName('datetime_field')).type(), QVariant.DateTime)
+
+        f = next(vl.getFeatures(QgsFeatureRequest()))
+
+        date_idx = vl.fields().lookupField('date_field')
+        self.assertIsInstance(f.attributes()[date_idx], QDate)
+        self.assertEqual(f.attributes()[date_idx], QDate(2004, 3, 4))
+        time_idx = vl.fields().lookupField('time_field')
+        self.assertIsInstance(f.attributes()[time_idx], QTime)
+        self.assertEqual(f.attributes()[time_idx], QTime(13, 41, 52))
+        datetime_idx = vl.fields().lookupField('datetime_field')
+        self.assertIsInstance(f.attributes()[datetime_idx], QDateTime)
+        self.assertEqual(f.attributes()[datetime_idx], QDateTime(QDate(2004, 3, 4), QTime(13, 41, 52)))
+
+    def testBinaryType(self):
+        create_sql = 'CREATE TABLE "qgis_test"."binary_type" ( ' \
+                     '"id" INTEGER NOT NULL PRIMARY KEY,' \
+                     '"blob" VARBINARY(114))'
+        insert_sql = 'INSERT INTO "qgis_test"."binary_type" ("id", "blob") VALUES (?, ?)'
+        insert_args = [[1, base64.b64encode(b'binvalue')], [2, None]]
```

Comment:
Why do we a base64 encoding when we'd like to test binary types? `b64encode` "converts" binary data to textual data. But I think we want to go the other way around when testing a binary type.

Maybe we can do `base64.b64decode('AAECAwQFBgcICQoLDA0ODw==')`. This will create a binary consisting of 16 bytes. The values go from 0 to 15.

Reply:
Done in 960bc2a.